Fix for pytest-xdist plugin compatibility

oiefimenko opened the pull request:
Fix for `pytest-xdist` plugin compatibility 

### Changes
`pytest-xdist` worker component provides additional assert of returning reports  relate to current scheduled test for execution. Thats  why original reruns failures works with it natively - it completes reruns just after tests failure before `runtest_protocol_complete` hook which is signal of end of test execution.

Current changes mimic such behavior. Also add reruns stats collection support for xdist launches such as each worker will generate separate stats with worker id prefix for filename.

### Testing
Local launches against synthetic test suite

Commit:
Added xdist support

Commit:
Comments added

lyapun reviewed (approved):
Tested locally and here - https://ci2.int.datarobot.com/job/paralleled_functional_master/19/

Looks like working 👍

Commit:
Added --max-tests-rerun param

oiefimenko commented:
@lyapun @andrew-qa @almiroshnich 
I added additional features:
* Reporting amount of time spend on reruns, ex:
```
=================Performed 2 reruns in 0.011191 seconds=================
```
* Added ability to provide threshold run failures value after which no reruns will be performed. Because of testrun sequence it will look a little bit strange since we need report failures one more time:
```
tests/test_tests_1.py::test_test_passing_1 PASSED                                                                                                                                                     [ 16%]
tests/test_tests_1.py::test_test_failing RERUN                                                                                                                                                        [ 33%]
tests/test_tests_1.py::test_test_passing_2 PASSED                                                                                                                                                     [ 50%]
tests/test_tests_2.py::test_test_passing_1 PASSED                                                                                                                                                     [ 66%]
tests/test_tests_2.py::test_test_failing RERUN                                                                                                                                                        [ 83%]
tests/test_tests_2.py::test_test_passing_2 PASSED                                                                                                                                                     [100%]
=============================================================== Too many failed test: 2 with threshold of 1. Restore failures without reruns ================================================================

tests/test_tests_1.py::test_test_failing FAILED                                                                                                                                                       [100%]
tests/test_tests_2.py::test_test_failing FAILED  
```

Please review

andrew-qa reviewed (approved):
Additional 👍 for threshold

lyapun reviewed (approved):
Cool!

> Added ability to provide threshold run failures value after which no reruns will be performed. 

What value we will set for this threshold?

oiefimenko commented:
@lyapun Idea is to set it to some value that will help us not to rerun large amount of failures during run. For example we have run with 30 valid failed tests - so not to retry all of them we could set threshold to 10. Definitely current approach is a lot empirical but we rarely have more then 10 flakes in a run.

lyapun commented:
Should not it be percentage instead?

E.g. if more than half of tests failed - we don't rerun anything. Something like this to prevent cases when all or almost all tests failing.

andrew-qa commented:
@lyapun Idea was to keep this numbers quite low, about 7-10 failures. To fail rather earlier then wait for some percentage of failed tests.

lyapun commented:
So is it amount of how many tests we allow to fail on rerun before we stop doing it? Or is it just amount of how many tests we gonna retry? Because if second - 7-10 failures might be too low amount if we have 1000 tests in batch, but too big amount if we have 10 tests in batch.

oiefimenko commented:
@lyapun Current behavior is: It is maximum amount of failed tests we will try to rerun. If amount of failures will exceed current threshold  plugin won't execute any reruns at all.

lyapun commented:
Right, so this is why I worried about that it will be hard to decide on the right value for this threshold

oiefimenko commented:
We could check all green jobs with reruns and check amount of reruns in them. And then come up with average or mix or etc

naspeh commented on pytest_rerunfailures.py:
```diff
@@ -81,6 +83,14 @@ def pytest_addoption(parser):
+        dest='max_tests_rerun',
+        type=int,
+        default=None,
+        help='provide path to export reruns artifact.'
```
help message for not this option :)

naspeh commented on pytest_rerunfailures.py:
```diff
@@ -424,9 +474,28 @@ def _save_reruns_artifact(self, session):
+    @contextmanager
+    def _prepare_xdist(self, item):
+        """
+        Explicetly changing current working test for xdist worker with rollback
```
*Explicitly

naspeh reviewed (approved):
So we still run all `rerun` in one process if xdist 2+ process, it's acceptable for sure.

But would be cool to run reruns in the same way as xdist (in the same amount of processes), but probably it would be hard to adjust %)

Some minor comments.

Commit:
PR comments address

oiefimenko commented:
@andrew-qa  @lyapun  @almiroshnich @naspeh 
I generated report on Kibana on amount of flakes resolved by reruns:
http://starscream.hq.datarobot.com:5601/app/kibana#/visualize/edit/Amount-of-flakes?_g=(filters:!(),refreshInterval:(display:Off,pause:!f,value:0),time:(from:now-60d,mode:quick,to:now))&_a=(filters:!((meta:(disabled:!f,index:jenkins,key:schema,negate:!f,value:rerun_test),query:(match:(schema:(query:rerun_test,type:phrase)))),(meta:(disabled:!f,index:jenkins,key:query,negate:!f,value:%27schema:rerun_test%27),query:(query_string:(analyze_wildcard:!t,query:%27schema:rerun_test%27))),(meta:(disabled:!f,index:jenkins,key:query,negate:!f,value:%27status:flake%27),query:(query_string:(analyze_wildcard:!t,query:%27status:flake%27)))),linked:!f,query:(query_string:(analyze_wildcard:!t,query:%27*%27)),vis:(aggs:!((id:%272%27,params:(),schema:metric,type:count),(id:%274%27,params:(field:build_number,order:desc,orderBy:%272%27,row:!f,size:500),schema:split,type:terms),(id:%275%27,params:(filters:!((input:(query:(query_string:(analyze_wildcard:!t,query:%27status:flake%27))),label:%27%27))),schema:group,type:filters)),listeners:(),params:(addLegend:!t,addTimeMarker:!f,addTooltip:!t,defaultYExtents:!f,mode:stacked,scale:linear,setYExtents:!f,shareYAxis:!t,spyPerPage:10,times:!(),yAxis:()),type:histogram))

But it is aggregated by master job, it is specific of collection of this data.
So I created small script and collect this data manually for 3000 last launches of `paralellel_function_executor` job:

```
Failures count  |  Jobs count   | %
    0           | 1246          | 41.5 %
    1           | 516           | 17.2 %
    2           | 289           | 9.6 %
    3           | 511           | 17.0 %
    4           | 173           | 5.7 %
    5           | 57            | 1.9 %
    6           | 35            | 1.2 %
    7           | 15            | 0.5 %
    8           | 101           | 3.4 %
    9           | 25            | 0.8 %
    10          | 1             | 0.03 %
    11          | 1             | 0.03 %
    35          | 1             | 0.03 %
    22          | 1             | 0.03 %
    23          | 1             | 0.03 %
```

lyapun commented:
So we will set it to `10`?

oiefimenko commented:
@lyapun I think it will be good option to start.
